nexus_motion_planner/src/motion_plan_cache.hpp

```diff
@@ -0,0 +1,165 @@
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <lifecycle_msgs/msg/state.hpp>
+
+// moveit modules
+#include <moveit/move_group_interface/move_group_interface.h>
+
+// NEXUS messages
+#include <nexus_endpoints.hpp>
+
+namespace nexus {
+namespace motion_planner {
+
+/**
+ * Cache manager for the Nexus motion planner.
+ *
+ * This manager facilitates cache management for MoveIt 2's `MoveGroupInterface`
+ * by using `warehouse_ros` to manage a database of executed motion plans, and
+ * how long they took to execute. This allows for the lookup and reuse of the
+ * best performing plans found so far.
+ *
+ * WARNING:
+ *   This cache does NOT support collision detection!
+ *   Plans will be put into and fetched from the cache IGNORING collision.
+ *   If your planning scene is expected to change between cache lookups, do NOT
+ *   use this cache, fetched plans are likely to result in collision then.
+ *
+ *   To handle collisions this class will need to hash the planning scene world
+ *   msg (after zeroing out std_msgs/Header timestamps and sequences) and do an
+ *   appropriate lookup.
+ *
+ * Relevant ROS Parameters:
+ *   - `warehouse_plugin`: What database to use
+ *   - `warehouse_host`: Where the database should be created
+ *   - `warehouse_port`: The port used for the database
+ *
+ * Motion plans are stored in the `move_group_plan_cache` database within the
+ * database file, with plans for each move group stored in a collection named
+ * after the relevant move group's name.
+ *
+ * For example, the "my_move_group" move group will have its cache stored in
+ * `move_group_plan_cache@my_move_group`
+ *
+ * Motion plans are keyed on:
+ *   - Plan Start: robot joint state
+ *   - Plan Goal (either of):
+ *     - Final pose (wrt. `planning_frame` (usually `base_link`))
+ *     - Final robot joint states
+ *
+ * Motion plans may be looked up with some tolerance at call time.
+ */
+class MotionPlanCache
+{
+public:
+  // We explicitly need a Node::SharedPtr because warehouse_ros ONLY supports
+  // it...
+  explicit MotionPlanCache(const rclcpp::Node::SharedPtr& node);
+
+  void init(
+    const std::string& db_path = ":memory:",
+    uint32_t db_port = 0,
+    double exact_match_precision = 1e-6);
+
+  // ===========================================================================
+  // MOTION PLAN CACHING
+  // ===========================================================================
+  // TOP LEVEL OPS
+  std::vector<
+    warehouse_ros::MessageWithMetadata<
+      moveit_msgs::msg::RobotTrajectory
+    >::ConstPtr
+  >
+  fetch_all_matching_plans(
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const std::string& move_group_namespace,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    double start_tolerance, double goal_tolerance, bool metadata_only = false);
+
+  warehouse_ros::MessageWithMetadata<
+    moveit_msgs::msg::RobotTrajectory
+  >::ConstPtr
+  fetch_best_matching_plan(
```

Comment:
Can you add docs explaining how the best plan is determined.

Reply:
Added comment, and even made it configurable now (defaults to `execution_time_s`, no behavior change.

6004f96fc39d4b287ab5c142adb1fc220c83a4b8 in https://github.com/osrf/nexus/pull/26